Add canonical URL for rustdoc pages

jsha opened the pull request:
Omit the link for crates that specify their own documentation URL.

Part of #74 and #1438

jyn514 commented on src/web/rustdoc.rs:
```diff
@@ -483,6 +484,24 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
+            .documentation_url
+            .as_ref()
+            .unwrap()
+            .starts_with("https://docs.rs/")
```
should we also do this for `http://docs.rs`?

jyn514 reviewed (commented):
This looks good to me, but can we add some tests to make sure it works properly?

jyn514 commented on src/web/rustdoc.rs:
```diff
@@ -483,6 +484,24 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
+            inner_path
+        )
+    } else {
+        "".to_string()
```
shouldn't this be the crate's `documentation_url`?
cc @Nemo157 I know you had concerns about this not actually being a rustdoc page

syphar replied:
since the canonical url should point to the _same page_, I don't see how this would work for `documentation_url` where we don't even know if it's generated with rustdoc, or just some mdbook.

syphar commented on src/web/rustdoc.rs:
```diff
@@ -483,6 +484,24 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
+        format!(
+            "{}://{}/{}/latest/{}",
+            req.url.scheme(),
+            req.url.host(),
```
canonical URLs can be anything that a link accepts, so we also could just omit everything except the path

jsha replied:
According to https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls:

> Use absolute paths rather than relative paths with the rel="canonical" link tag.
> Recommended: https://www.example.com/dresses/green/greendress.html
> Not recommended: /dresses/green/greendress.html

syphar replied:
ok, thanks for the info, didn't know that!

jyn514 replied:
not sure I heard an answer here - @jsha why did you choose to use `req.url.host()`?

syphar commented on templates/rustdoc/head.html:
```diff
@@ -3,4 +3,8 @@
         <link rel="search" href="/-/static/opensearch.xml" type="application/opensearchdescription+xml" title="Docs.rs" />
 
+        {%- if canonical_url -%}
+        <link rel="canonical" href="{{canonical_url | safe}}" />
```
@jyn514 wouldn't we need the same logic for other versioned pages? The docs.rs crate page comes to mind

jsha replied:
IMO we shouldn't bother. Almost no one searches for the crate page, so it's not that critical to add canonicalization for it. If we find down the line that people are looking for that page more often, we can always add canonicalization later.

syphar commented on src/web/rustdoc.rs:
```diff
@@ -483,6 +484,22 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
+        format!(
+            "https://docs.rs/{}/latest/{}",
+            name,
+            inner_path
```
there is no guarantee that a page that exists in one version also exists in another version. 

That's why we use the `target-redirect` handler for the links in the dropdown. 

Should we put a link to the target-redirect here into the canonical url? 

Or replicate part of the logic in `target_redirect_handler` to show something more usable here? 

( since archive storage file-exist queries are relatively fast and local, but we probably should measure)

syphar replied:
Ah, understood, this is probably worth a comment so nobody is confused. 

at some point we could optimize this by pointing to the specific latest version where this page exists, but this would need faster file-exist checks.

syphar commented:
@jsha I assume the second part of https://github.com/rust-lang/docs.rs/issues/74#issuecomment-1184900299 will be in a second PR? So `no-index` for pages which provide a documentation url?

jsha commented:
> I assume the second part of https://github.com/rust-lang/docs.rs/issues/74#issuecomment-1184900299 will be in a second PR? So no-index for pages which provide a documentation url?

Yes, that's correct. That part is a little trickier - I'll update on #74.

jsha commented:
Pushed a new revision with a test. Writing a test helped me notice we were including `index.html` unnecessarily in the canonical URLs; I removed that.

Added a comment describing why it's fine that the target might in some cases not exist.

jyn514 reviewed (approved):
Lgtm other than the nit. Thanks!

jsha commented:
Sure! I named it after the corresponding field in Metadata. Should I change
that too?

jsha commented on src/web/rustdoc.rs:
```diff
@@ -1980,4 +2003,64 @@ mod test {
+            env.fake_release()
+                .name("dummy-dash")
+                .version("0.1.0")
+                .documentation(Some("http://example.com".to_string()))
```
Done! I wound up not renaming the field in Metadata since it's meant to match what's in Cargo.toml.
